File: Array/219_ContainsDuplicateII.py
from typing import List


class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        pos = {}
        for i, num in enumerate(nums):
            if num in pos and i - pos[num] <= k:
                return True
            pos[num] = i
        return False


# sliding window
class Solution2:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        win = set()
        for i, num in enumerate(nums):
            if i > k:
                win.remove(nums[i - k - 1])
            if num in win:
                return True
            win.add(num)
        return False

File: Java Solutions/219_ContainsDuplicateII.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        HashMap<Integer, Integer> pos = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (pos.containsKey(nums[i]) && i - pos.get(nums[i]) <= k) {
                return true;
            }
            pos.put(nums[i], i);
        }
        return false;
    }
}


class Solution2 {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        HashSet<Integer> seen = new HashSet<>();
        for (int i = 0; i < nums.length; i++) {
            if (seen.contains(nums[i])) return true;
            seen.add(nums[i]);
            if (seen.size() > k) {
                seen.remove(nums[i - k]);
            }
        }
        return false;
    }
}